src/internal/tokencache/default_storage_manager_test.go

```diff
@@ -311,74 +312,203 @@ func TestWriteIDToken(t *testing.T) {
+				familyID:      "fid",
+				clientID:      "cid",
+			},
+			want: testRefreshTokenWithFID,
+		},
+		{
+			name: "Token with fid, read with cid, env, and hid",
+			m: &defaultStorageManager{
+				refreshTokens: map[string]*refreshTokenCacheItem{
+					testRefreshTokenWoFID.CreateKey(): testRefreshTokenWithFID,
+				},
+			},
+			args: args{
+				homeAccountID: "hid",
+				envAliases:    []string{"test", "env", "hello"},
+				familyID:      "",
+				clientID:      "cid",
+			},
+			want: testRefreshTokenWithFID,
+		},
+		{
+			name: "Token with fid, verify CID is not required", // match on hid, env, and has fid
+			m: &defaultStorageManager{
+				refreshTokens: map[string]*refreshTokenCacheItem{
+					testRefreshTokenWoFID.CreateKey(): testRefreshTokenWithFID,
+				},
+			},
+			args: args{
+				homeAccountID: "hid",
+				envAliases:    []string{"test", "env", "hello"},
+				familyID:      "",
+				clientID:      "",
+			},
+			want: testRefreshTokenWithFID,
+		},
+		{
+			name: "Token with fid, Verify env is required",
+			m: &defaultStorageManager{
+				refreshTokens: map[string]*refreshTokenCacheItem{
+					testRefreshTokenWoFID.CreateKey(): testRefreshTokenWithFID,
+				},
+			},
+			args: args{
+				homeAccountID: "hid",
+				envAliases:    []string{},
+				familyID:      "",
+				clientID:      "",
+			},
+			want: nil,
+		},
+		{
+			name: "Multiple items in cache, given a fid, item with fid will be returned",
+			m: &defaultStorageManager{
+				refreshTokens: map[string]*refreshTokenCacheItem{
+					testRefreshTokenWoFID.CreateKey():       testRefreshTokenWoFID,
+					testRefreshTokenWithFID.CreateKey():     testRefreshTokenWithFID,
+					testRefreshTokenWoFIDAltCID.CreateKey(): testRefreshTokenWoFIDAltCID,
+				},
+			},
+			args: args{
+				homeAccountID: "hid",
+				envAliases:    []string{},
+				familyID:      "fid",
+				clientID:      "cid",
+			},
+			want: nil,
+		},
+		// Cannot guarentee that without an alternate cid which token will be
+		// returned deterministically when HID, CID, and env match.
+		{
+			name: "Multiple items in cache, without a fid and with alternate CID, token with alternate CID is returned",
+			m: &defaultStorageManager{
+				refreshTokens: map[string]*refreshTokenCacheItem{
+					testRefreshTokenWoFID.CreateKey():       testRefreshTokenWoFID,
+					testRefreshTokenWithFID.CreateKey():     testRefreshTokenWithFID,
+					testRefreshTokenWoFIDAltCID.CreateKey(): testRefreshTokenWoFIDAltCID,
+				},
+			},
+			args: args{
+				homeAccountID: "hid",
+				envAliases:    []string{},
+				familyID:      "",
+				clientID:      "cid2",
+			},
+			want: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.m.ReadRefreshToken(tt.args.homeAccountID, tt.args.envAliases, tt.args.familyID, tt.args.clientID); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("defaultStorageManager.ReadRefreshToken() = %v, want %v", render.AsCode(got), render.AsCode(tt.want))
```

Comment:
bringing in render as a test dependency so we get meaningful test failures. with just using %v on a struct with a pointer, you only get the pointer value (an address)